### PR TITLE
Fix S3 Storage Initialiser downloading additional files when given file path does not include file extension

### LINKED ISSUE #2473: S3 storage initializer doesn't work as expected for file without extension
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

I created the following files in my s3 bucket:
```
test/churn-pickle
test/churn-pickle-logs
test/churn-pickle-report
```
`churn-pickle`, `churn-pickle-logs` and `churn-pickle-report` are `file` without `extension`.

I use storage initializer to download `s3://<bucket>/test/churn-pickle` to `/mnt/models`. However, I got the following result:

```
❯ ls /mnt/models
total 24
-rw-r--r--  1 i543026  staff    12B 14 Oct 19:25 -logs
-rw-r--r--  1 i543026  staff    12B 14 Oct 19:25 -report
-rw-r--r--  1 i543026  staff    12B 14 Oct 19:25 churn-pickle
```

**What did you expect to happen:**

1. if the given path is a `file`, only the file is downloaded. In this case, `s3://<bucket>/test/churn-pickle` will be downloaded to `/mnt/models/churn-pickle`.
2. If the given path is a `folder`, all the files under the folder will be downloaded. 
3. If the given path is a `prefix` (not a folder or file), nothing should be downloaded.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

I assume that the storage initializer can only download from folders and files, not from `prefix`.

The issue was probably caused by:
1. objects are filter by the [prefix](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L167).
2. replace logic [here](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L167).

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version: 0.9
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`): v1.21.14
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by tjandy98**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** :
Fixes #2473

**Type of changes**
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: python/kserve/kserve/storage.py [hunk 1]
```diff
@@ -181,7 +181,7 @@ def _download_s3(uri, temp_dir: str):
             target_key = (
                 obj.key.rsplit("/", 1)[-1]
                 if bucket_path == obj.key
-                else obj.key.replace(bucket_path, "", 1).lstrip("/")
+                else obj.key.replace(bucket_path.rsplit("/",1)[0], "", 1).lstrip("/")
```
Comment (yuzisun) [inline python/kserve/kserve/storage.py]: Can you help add a test ? Thanks !

Comment (tjandy98) [reply python/kserve/kserve/storage.py]: @yuzisun I have added it, thank you.

Comment (yuzisun) [reply python/kserve/kserve/storage.py]: Thanks ! there is a small lint issue and please also help update the PR with a meaningful title as that goes into the release notes.

File: python/kserve/kserve/storage.py [hunk 2]
```diff
@@ -181,7 +181,7 @@ def _download_s3(uri, temp_dir: str):
             target_key = (
                 obj.key.rsplit("/", 1)[-1]
                 if bucket_path == obj.key
-                else obj.key.replace(bucket_path, "", 1).lstrip("/")
+                else obj.key.replace(bucket_path.rsplit("/", 1)[0], "", 1).lstrip("/")
```
Comment (yuzisun) [inline python/kserve/kserve/storage.py]: @tjandy98 would this break the case `s3://test-bucket/a` that bucket path `test-bucket/a` becomes `test-bucket` after `rsplit` and object key `test-bucket/a/b/model.bin` is replaced with `dir/a/b/model.bin` which is supposed to be `dir/b/model.bin?

Comment (yuzisun) [reply python/kserve/kserve/storage.py]: >>> s = "test-bucket/a/b/model.bin"
>>> bucket_path = "test-bucket/a"
>>> s.replace(bucket_path.rsplit("/", 1)[0], "", 1).lstrip("/")
'a/b/model.bin'
>>> s.replace(bucket_path, "", 1).lstrip("/")
'b/model.bin'

Comment (tjandy98) [reply python/kserve/kserve/storage.py]: @yuzisun You are right, thanks for providing an example! I have made some changes and added a test based on this.

**Comment by yuzisun** [comment]
fixed in https://github.com/kserve/kserve/pull/3088

**Pull request closed without merging**